src/data_structures/suffix_array.rs

```diff
@@ -19,12 +20,64 @@ use vec_map::VecMap;
 
 use alphabets::{Alphabet, RankTransform};
 use data_structures::smallints::SmallInts;
+use data_structures::bwt::{BWT, Less, Occ};
 
+pub trait SuffixArray {
+    fn get(&self, index: usize) -> Option<usize>;
+    fn len(&self) -> usize;
+}
 
-pub type SuffixArray = Vec<usize>;
-pub type SuffixArraySlice = [usize];
+pub type RawSuffixArray = Vec<usize>;
+pub struct SampledSuffixArray<DBWT: Deref<Target = BWT>, DLess: Deref<Target = Less>, DOcc: Deref<Target = Occ>> {
+    bwt: DBWT,
+    less: DLess,
+    occ: DOcc,
+    sample: Vec<usize>,
+    s: usize, // Rate of sampling
+}
 pub type LCPArray = SmallInts<i8, isize>;
 
+impl SuffixArray for RawSuffixArray {
+    fn get(&self, index: usize) -> Option<usize> {
+        // Explicitly written out because Vec::get(index) generates a recursion warning
+        if index < self.len() {
+            Some(self[index])
+        } else {
+            None
+        }
+    }
+    fn len(&self) -> usize {
+        Vec::len(self)
+    }
+}
+
+impl<DBWT: Deref<Target = BWT>, DLess: Deref<Target = Less>, DOcc: Deref<Target = Occ>> SuffixArray for SampledSuffixArray<DBWT, DLess, DOcc> {
+    fn get(&self, index: usize) -> Option<usize> {
+        if index < self.len() {
+            let mut pos = index;
+            let mut offset = 0;
+            println!("pos sample: {:?}", self.sample);
```

Comment:
Sorry for the drive-by -- just excited to see the work here!

Is there a reason for the calls to `println!`?